src/Shared/TypeLoader.cs

```diff
@@ -345,79 +312,55 @@ internal LoadedType GetLoadedTypeByTypeName(string typeName)
             {
                 ErrorUtilities.VerifyThrowArgumentNull(typeName, "typeName");
 
-                // Only one thread should be doing operations on this instance of the object at a time.
-                lock (_lockObject)
+                if (Interlocked.Read(ref _haveScannedPublicTypes) == 0)
```

Comment:
Aren't there better/accepted ways to implement the singleton pattern in C#?  Lazy<T> was only added in .NET 4, but we're already adding a dependency on that, no?